### ISSUE TITLE
Mimimal compiler version

### ISSUE THREAD

**Issue by etam** [pioneerspacesim/pioneer#5810]
In COMPILING.txt it says

> Pioneer is known to build on the following platforms and build systems:
> Linux: CMake with GCC 4.9+ or Clang 6+

The commit eada00c0d8a008be6b791c8e75c8f18aabcb8f89 uses [`std::reduce`](https://en.cppreference.com/w/cpp/algorithm/reduce), which comes with c++17. This raises the required minimal compiler version, and requires adding a compiler flag.

**Comment by Gliese852** [comment]
Oh, apparently this is outdated information, we have been in C++17 for quite a long time

https://github.com/pioneerspacesim/pioneer/blob/c9abf3c3500f7c467083700f261da357e680426e/CMakeLists.txt#L5

Interesting that you had no problems for example with `std::variant`:

https://github.com/pioneerspacesim/pioneer/blob/c9abf3c3500f7c467083700f261da357e680426e/src/core/Property.h#L52

or with `std::string_view`:

https://github.com/pioneerspacesim/pioneer/blob/c9abf3c3500f7c467083700f261da357e680426e/src/core/StringUtils.cpp#L188

**Comment by etam** [comment]
I don't know if I have no problem with those other things. The compilation on the older system stops at the std::reduce and doesn't go any further.
It's not a problem for me and I'm not asking for compatibility.
It just means that COMPILING.txt needs a bit of an update.

**Comment by Gliese852** [comment]
Yeah, I just wanted to point out that commit eada00c does not raise the required minimal compiler version.

**Comment by impaktor** [comment]
So we should bump minimal GCC version from 4.9 to 9+?
And possibly clang from 6+ to 9+?

**Comment by Web-eWorks** [comment]
Clang 6 has 99% of C++17 support already present. All filed Defect Reports against C++17 are fully implemented as of Clang 9, though the only fixes which are not present in Clang 7 are to `[[nodiscard]]` constructors and structured bindings, which we do not make use of in the Pioneer codebase.

That being said, Clang 9 is ~4 years old at this point and can be expected to be present on most LTS distributions (e.g Ubuntu 20.04 uses Clang 10 as the default `clang` package). Feel free to bump the version. GCC 9 is also a good minimum base, as it is the first release with a stable ABI for C++17 (according to the GCC documentation).
